wave_lang/kernel/wave/water.py

```diff
@@ -446,3 +446,271 @@ def add_opt(pipeline):
+
+        for test_path in test_paths:
+            try:
+                debug_info.append(f"Testing import with path: {test_path}")
+                sys.path.insert(0, str(test_path))
+                import water_mlir
+                import water_mlir.ir
+
+                # If we get here, the import worked!
+                debug_info.append(f"✓ Import test successful with path: {test_path}")
+                return str(test_path)
+            except ImportError as e:
+                debug_info.append(f"  Import failed: {e}")
+                continue
+            finally:
+                sys.path[:] = original_path
+
+    except Exception as e:
+        debug_info.append(f"Final fallback failed: {e}")
+
+    # Print debug info if we can't find anything (for CI debugging)
+    debug_str = "\n".join(debug_info)
+    print(f"[DEBUG] Water Python path search failed. Debug info:\n{debug_str}")
+
+    return None
+
+
+def apply_water_passes_with_passmanager(mlir_text: str) -> str:
+    """Apply Water Wave passes using Python PassManager instead of subprocess.
+
+    This function applies the standard Wave lowering passes:
+    - water-wave-detect-normal-forms
+    - water-wave-propagate-elements-per-thread
+    - lower-wave-to-mlir
+    - canonicalize
+    - cse
+
+    Args:
+        mlir_text: Input Wave dialect MLIR as string
+
+    Returns:
+        Optimized MLIR as string after applying the passes
+
+    Raises:
+        RuntimeError: If Water Python bindings are not available or passes fail
+    """
+    import sys
+
+    # Find Water Python bindings path dynamically
+    water_python_path_str = _find_water_python_path()
+    if not water_python_path_str:
+        raise RuntimeError("Could not find Water Python bindings directory")
+
+    # Add to sys.path if not already present
+    if water_python_path_str not in sys.path:
+        sys.path.insert(0, water_python_path_str)
+
+    try:
+        # Lazy imports - only import when function is called
+        from water_mlir import ir as water_ir
+        from water_mlir import passmanager
+        from water_mlir.dialects import wave as wave_dialect
+    except ImportError as e:
+        # Remove the path we added to avoid polluting sys.path
+        if water_python_path_str in sys.path:
+            sys.path.remove(water_python_path_str)
+        raise RuntimeError(f"Water Python bindings not available: {e}")
+
+    try:
+        # Initialize MLIR context and register Wave dialect
+        with water_ir.Context() as ctx:
+            ctx.allow_unregistered_dialects = True
+            wave_dialect.register_dialect(ctx)
+
+            # Register Wave passes
+            wave_dialect.register_passes()
+
+            # Parse the input MLIR
+            module = water_ir.Module.parse(mlir_text, ctx)
+
+            # Apply the pass pipeline
+            pipeline_str = "builtin.module(water-wave-detect-normal-forms,water-wave-propagate-elements-per-thread,lower-wave-to-mlir,canonicalize,cse)"
```

Comment:
There's some logic to stich these together in a nicer way, in `make_linear_pass_pipeline`, that makes this look nicer.